Add `--build-output-path` argument to `tuist build`

luispadron opened the pull request:
Resolves https://github.com/tuist/tuist/issues/2392

### Short description 📝

This PR adds a new parameter to `tuist build`: 

`--build-output-path` which is intended to allow users to specify the path to a directory where the `xcodebuild` products will be copied to after `tuist build`.

Example:

Running `tuist build clide --build-output-path clide-build` produces:

<img width="257" alt="Screen Shot 2021-04-28 at 7 41 50 PM" src="https://user-images.githubusercontent.com/13840545/116485674-c83e1680-a859-11eb-8c41-6a46aa8d4371.png">

Running `tuist build clide --build-output-path clide-build --configuration Release` after the above produces:

<img width="393" alt="Screen Shot 2021-04-28 at 7 43 09 PM" src="https://user-images.githubusercontent.com/13840545/116485749-f91e4b80-a859-11eb-869c-1f0cc53782cd.png">

## Background

When running `xcodebuild` command, Xcode will place build products to (by default): `~/Library/Developer/DerivedData/PROJECT_NAME-HASH/Build/Products/CONFIGURATION_NAME`

Since this path is generated by Xcode we would like a way to configure where the build products go. With `--build-output-path` you can specify a location where `~/Library/Developer/DerivedData/PROJECT_NAME-HASH/Build/Products` will be copied to.

The directory will contain other directories (based on configuration) where the build products for that configuration will be copied to.

### Checklist ✅

- [X] The code architecture and patterns are consistent with the rest of the codebase.
- [X] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [X] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [X] In case the PR introduces changes that affect users, the documentation has been updated.

pepibumur commented on Package.resolved:
```diff
@@ -132,8 +132,8 @@
         "repositoryURL": "https://github.com/stencilproject/Stencil.git",
         "state": {
           "branch": null,
-          "revision": "94197b3adbbf926348ad8765476a158aa4e54f8a",
```
I've seen this bumped in another PR too. I'd remove this from this PR and open a separate PR for the version bump. That way we can capture the changes in the CHANGELOG and make it easier to revert in case it causes issues.

fortmarek replied:
Yeah, sometimes SPM updates the packages just by you opening a project via Xcode (or maybe Xcode does it). It's quite annoying

luispadron replied:
For sure, i think this was Xcode. Will undo 👍🏼

pepibumur commented on Sources/TuistCore/Utils/XcodeBuildLocator.swift:
```diff
@@ -0,0 +1,77 @@
+import TuistGraph
+import TuistSupport
+
+public protocol XcodeBuildLocating {
```
Nitpick: What about renaming this one to `XcodeProjectBuildDirectoryLocating`. It's a bit longer but more explicit. I think it'll rule out any future confusion about what the utility is for.

pepibumur replied:
If you agree with renaming I'd do it with all the definitions in this file, including the error.

luispadron replied:
Yeah considered this but kept the name that @natanrolnik had used but I like the explicit name more.

pepibumur commented on Sources/TuistCore/Utils/XcodeBuildLocator.swift:
```diff
@@ -0,0 +1,77 @@
+    ///
+    /// - Parameters:
+    ///   - projectPath: The path of the Xcode project or workspace.
+    func locateProductsDirectory(for projectPath: AbsolutePath) throws -> AbsolutePath
```
If we rename the class, its name will already be explicit about what we are locating so we can rename this one to just `locate` and remove some naming redundancy.

pepibumur commented on Sources/TuistCore/Utils/XcodeBuildLocator.swift:
```diff
@@ -0,0 +1,77 @@
+            if platform == .macOS {
+                return configuration
+            } else {
+                guard let sdk = platform.xcodeSimulatorSDK else {
```
Since `macOS` is the only platform without simulators, I think we can remove the error handling her and do `platform.xcodeSimulatorSDK!`. It's very unlikely that this assumption will change and therefore cause a runtime error here so we can simplify this file a little bit by removing the error and force-unwrapping.

pepibumur commented on Sources/TuistCore/Utils/XcodeBuildLocator.swift:
```diff
@@ -0,0 +1,77 @@
+    ///   - platform: The platform for the built scheme.
+    ///   - projectPath: The path of the Xcode project or workspace.
+    ///   - configuration: The configuration name, i.e. `Release`, `Debug`, or something custom.
+    func locateBuildDirectory(
```
I'd unit test this method:
- Returns the right path when the platform is macOS
- Returns the right path when the platform is not macOS

pepibumur commented on Sources/TuistKit/Services/BuildService.swift:
```diff
@@ -70,41 +77,87 @@ final class BuildService {
+        }
+
+        logger.log(level: .notice, "Copying build products to \(outputPath.pathString)", metadata: .subsection)
+        try FileHandler.shared.copy(from: buildOutputDirectory, to: outputPath)
```
Am I right assuming that we might copy files from previous builds? Here are some ideas:
- We can get a timestamp before building, and recursively copy the files whose updated at date is more recent than the timestamp.
- Since we build a configuration, you can copy he directory for only that configuration.

Also, as a user, let's say I have the following under the `--products-path` directory:

```
products/
  Debug/
```

If I build for `Release`, I'd expect to end up with something like this:

```
products/
  Debug/
  Release/
```
Note that we are not overriding previous configurations. This behavior is similar to the SPM's.

luispadron replied:
> Am I right assuming that we might copy files from previous builds?

We delete the directory for the current build configuration so in this example if the path was `.build` and building `Release` configuration, the implementation right now deletes the directory `.build/Release` and then copies from derived data to the cleaned `.build/Release`.

> Note that we are not overriding previous configurations. This behavior is similar to the SPM's.

For sure! We do are doing this here as well since we copy to a directory based on configuration. So `--build-path .build --configuration Release` -> `.build/Release` getting copied

luispadron replied:
Checkout comments below to see if we should update this to also include the scheme name in the products directory.

pepibumur reviewed (changes requested):
This is awesome @luispadron 👏. Besides the comments that I left, I'd update the documentation to document the new argument, and also add an acceptance tests along the lines of:
- I generate the project.
- I build the project with the products path X
- I expect the directory at X to contain files foo, bar..

fortmarek commented on Sources/TuistKit/Services/BuildService.swift:
```diff
@@ -70,41 +77,87 @@ final class BuildService {
+        let buildOutputDirectory = try xcodeBuildLocator.locateProductsDirectory(for: workspacePath)
+
+        guard FileHandler.shared.exists(buildOutputDirectory) else {
+            logger.error("No build products found at \(buildOutputDirectory.pathString), skipping copy products")
```
I don't think it's completely right to log an error but not throw an error - I think we can throw an error here instead

luispadron replied:
Agree! Throwing a `.bug` here.

fortmarek commented:
Thanks for doing this! This will be much appreciated. One thing I am wondering is whether we need to let users specify the output directory - because alternatively, we could do the copy every time to e.g. `.tuist-outputs` directory (which would be documented and should be in `.gitignore`) and users would not have to really specify anything. But maybe there are valid use cases for people to configure that directory to what they want

luispadron commented:
@fortmarek I agree we should make this the default! I think as we've seen with the cache path it's also useful to be able to configure. 

Idea:

By default Tuist should store all build products under:

`~/.tuist/Builds/ProductName/ConfigurationName`

We'd document this and users can use this as a sane default over the weirdness of Xcode derived data. We should also be careful to make sure we don't use the build / set expectations this is cleared when clearing derived data in Xcode though!

Additionally users can define this location via `--build-path` flag. 

Side note:

I’m back and forth on the flag name, i think `--build-path` makes sense to me when we consider it's called with `tuist build` but maybe it's confusing?

Either way let me know what you all think about the default support + config option.

fortmarek commented:
> I think as we've seen with the cache path it's also useful to be able to configure.

+1, this is mostly good for CI where you generally don't want to work with global state

> By default Tuist should store all build products under:

👍  I'd maybe consider outputting the path after a build succeeds apart from adding to the documentation as well, so it's easier to find.

> I’m back and forth on the flag name, i think --build-path makes sense to me when we consider it's called with tuist build but maybe it's confusing?

Yeah, maybe something more specific like `--build-output-path`? I agree `--build-path` could be a little bit confusing.

luispadron commented:
@pepibumur @fortmarek 

Thanks for the review! I've addressed most of the comments in the PR. 

I'm going to defer work on the `~/.tuist/Builds` being the default location where build products get copied, it will be tracked by: #2870. I'd like to discuss how this should work with the cache / clean, I also noticed we have a `BuildCache` directory already so would like to talk about this implementation more and rather not tie it with this PR.

As for the current implementation, would like some thoughts on this:

When running `tuist build MyScheme --build-output-path .build --configuration debug` currently this will copy products to: `.build/debug`. But if we then run `tuist build MyOtherScheme --build-output-path .build --configuration debug` we would delete the existing `.build/debug` since the directory is named after configuration. 

Do we see this being problem? If so, we could instead copy to `.build/MyScheme/debug` and `.build/MyOtherScheme/debug` this way you get to build different schemes without overwriting the existing builds.

luispadron commented:
Update, I've updated the logic for copying as follow:

When Xcode builds products to: `~/Library/Developer/Xcode/DerivedData/MyProject-Hash/debug-iphonesimulator` it doesn't do so based on a scheme. This means multiple products are in the `debug-iphonesimulator` directory. Previously, in this PR we were starting a fresh copy (deleting the output-path every build), instead now we copy over **all** the products from the Xcode build path and only remove items that already exist. The rest of the items / directories are not touched. This also makes it compatible with `.build` directory from SPM.

In the current updated PR we copy each individual product over, this means that if you were to run the command:
`tuist build App --configuration debug --output-path build` you would get a result similar to:

```
build
├── debug-iphonesimulator
│   ├── App.app
│   ├── App.swiftmodule
│   └── FrameworkA.framework
```

When re-running this command with `tuist build MyOtherApp --configuration debug --output-path build`, you will see that we don't overwrite the entire folder:

```
build
├── debug-iphonesimulator
│   ├── App.app
│   ├── App.swiftmodule
│   ├── MyOtherApp.app
│   ├── MyOtherApp.swiftmodule
│   └── FrameworkA.framework
```

And finally when running a different configuration, say: `tuist build App --configuration --output-path build` the result would look like:

```
projects/tuist/fixtures/ios_app_with_custom_configuration/clide-build/
├── debug-iphonesimulator
│   ├── App.app
│   ├── App.swiftmodule
│   ├── MyOtherApp.app
│   ├── MyOtherApp.swiftmodule
│   └── FrameworkA.framework
└── release-iphonesimulator
    ├── App.app
    ├── App.app.dSYM
    ├── App.swiftmodule
    ├── FrameworkA.framework
    └── FrameworkA.framework.dSYM
```

I believe this behavior is the most standard one of the different attempts I tried. It also plays nicely with other build tools like SPM. Let me know what you think. I've added more tests as well.

pepibumur reviewed (approved):
Thanks a lot for addressing the comments @luispadron 🙏. One tiny request I'd make and I think we can then merge is renaming it to `--build-output-path` as @fortmarek suggested.

fortmarek commented on CHANGELOG.md:
```diff
@@ -217,7 +218,7 @@ Please, check out guidelines: https://keepachangelog.com/en/1.0.0/
 ### Changed
 
-- The parameter `--path` of `tuist graph` now specifies where the manifest is. To specify the output directory of the graph, use `--output-path` [#2434]((https://github.com/tuist/tuist/pull/2434) by [@adellibovi](https://github.com/adellibovi).
+- The parameter `--path` of `tuist graph` now specifies where the manifest is. To specify the output directory of the graph, use `--build-output-path` [#2434]((https://github.com/tuist/tuist/pull/2434) by [@adellibovi](https://github.com/adellibovi).
```
I believe this is not correct?

luispadron replied:
oops, i replaced `--output-path` so it did this too lol, going to undo

fortmarek commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ Please, check out guidelines: https://keepachangelog.com/en/1.0.0/
 ### Added
 
 - Add `tuist dependencies update` command. [#2819](https://github.com/tuist/tuist/pull/2819) by [@laxmorek](https://github.com/laxmorek)
+- Add `--build-output-path` option to `tuist build` by [@Luis Padron](https://github.com/luispadron).
```
missing link to your PR 🙂

fortmarek commented on Tests/TuistCoreTests/Utils/XcodeProjectBuildDirectoryLocatorTests.swift:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    override func tearDown() {
+        super.tearDown()
```
this should be run as the last statement in this function

fortmarek reviewed (approved):
Thanks @luispadron for revising the copy mechanism, I think it's perfect now 👏  apart from CI needs, this will also play very nicely with `Tasks` 👀  Will try to make more progress on that this week
